pkg/controller/servicemesh/controlplane/readiness.go

```diff
@@ -219,36 +232,76 @@ func (r *controlPlaneInstanceReconciler) isCNIReady(ctx context.Context) (bool,
+func (r *controlPlaneInstanceReconciler) calculateReadinessForType(ctx context.Context, namespaces []string, list runtime.Object, isReady isReadyFunc, readinessMap map[string]bool) error {
 	log := common.LogFromContext(ctx)
 
-	meshNamespace := r.Instance.GetNamespace()
-	selector := map[string]string{common.OwnerKey: meshNamespace}
-	err := r.Client.List(ctx, list, client.InNamespace(meshNamespace), client.MatchingLabels(selector))
-	if err != nil {
-		return err
-	}
-	items, err := meta.ExtractList(list)
-	if err != nil {
-		return err
-	}
-
-	for _, obj := range items {
-		metaObject, err := meta.Accessor(obj)
+	selector := map[string]string{common.OwnerKey: r.Instance.GetNamespace()}
+	for _, ns := range namespaces {
+		err := r.Client.List(ctx, list, client.InNamespace(ns), client.MatchingLabels(selector))
 		if err != nil {
 			return err
 		}
-		if component, ok := metaObject.GetLabels()[common.KubernetesAppComponentKey]; ok {
-			ready, exists := readinessMap[component]
-			readinessMap[component] = (ready || !exists) && isReady(obj)
-		} else {
-			// how do we have an owned resource with no component label?
-			log.Error(nil, "skipping resource for readiness check: resource has no component label", obj.GetObjectKind(), metaObject.GetName())
+		items, err := meta.ExtractList(list)
+		if err != nil {
+			return err
+		}
+
+		for _, obj := range items {
+			log.V(3).Info("Readiness check found object", "object", obj)
+			metaObject, err := meta.Accessor(obj)
+			if err != nil {
+				return err
+			}
+			if component, ok := metaObject.GetLabels()[common.KubernetesAppComponentKey]; ok {
+				ready, exists := readinessMap[component]
+				readinessMap[component] = (ready || !exists) && isReady(obj)
+			} else {
+				// resource was most likely created by user, not by the operator; we can safely ignore it
+				log.V(3).Info("skipping resource for readiness check: resource has no component label", obj.GetObjectKind(), metaObject.GetName())
+			}
 		}
 	}
 	return nil
 }
 
+func (r *controlPlaneInstanceReconciler) getNamespacesToCheck() ([]string, error) {
+	// gather namespaces in SMCP
+	namespaces := sets.NewString(r.Instance.Namespace)
+	if gateways := r.Instance.Spec.Gateways; gateways != nil {
+		if gw := gateways.ClusterIngress; gw != nil {
+			if gw.Namespace != "" {
+				namespaces.Insert(gw.Namespace)
+			}
+		}
+		if gw := gateways.ClusterEgress; gw != nil {
+			if gw.Namespace != "" {
+				namespaces.Insert(gw.Namespace)
+			}
+		}
+		for _, gw := range gateways.IngressGateways {
+			if gw.Namespace != "" {
+				namespaces.Insert(gw.Namespace)
+			}
+		}
+		for _, gw := range gateways.EgressGateways {
+			if gw.Namespace != "" {
+				namespaces.Insert(gw.Namespace)
+			}
+		}
+	}
+
+	// ensure we only check namespaces that are part of the mesh
+	smmr := &maistrav1.ServiceMeshMemberRoll{}
+	if err := r.Client.Get(context.TODO(), client.ObjectKey{Namespace: r.Instance.Namespace, Name: common.MemberRollName}, smmr); err != nil {
```

Comment:
do we have an informer/cache to use instead of making an api call?

Reply:
Internally, I think the controller-runtime client delegates to a cache, as long as you're not using unstructured object types.

Reply:
Yes, this hits the cache.